src/graph/sampler.cc

```diff
@@ -991,6 +991,118 @@ Subgraph NegEdgeSubgraph(int64_t num_tot_nodes, const Subgraph &pos_subg,
+  int64_t last_chunk_size = num_pos_edges - num_chunks * chunk_size;
+  int64_t num_neg_edges = (chunk_size * 2 - 1) * chunk_size * 2 * num_chunks;
+  int64_t num_neg_edges_last_chunk = (last_chunk_size * 2) * (last_chunk_size - 1) * 2;
+  int64_t num_all_neg_edges = num_neg_edges + num_neg_edges_last_chunk;
+  IdArray neg_dst = IdArray::Empty({num_all_neg_edges}, coo->dtype, coo->ctx);
+  IdArray neg_src = IdArray::Empty({num_all_neg_edges}, coo->dtype, coo->ctx);
+  IdArray neg_eid = IdArray::Empty({num_all_neg_edges}, coo->dtype, coo->ctx);
+  IdArray induced_neg_eid = IdArray::Empty({num_all_neg_edges}, coo->dtype, coo->ctx);
+
+  // These are vids in the positive subgraph.
+  const dgl_id_t *dst_data = static_cast<const dgl_id_t *>(coo->data);
+  const dgl_id_t *src_data = static_cast<const dgl_id_t *>(coo->data) + num_pos_edges;
+  const dgl_id_t *induced_vid_data = static_cast<const dgl_id_t *>(pos_subg.induced_vertices->data);
+  const dgl_id_t *induced_eid_data = static_cast<const dgl_id_t *>(pos_subg.induced_edges->data);
+  size_t num_pos_nodes = pos_subg.graph->NumVertices();
+  std::vector<size_t> pos_nodes(induced_vid_data, induced_vid_data + num_pos_nodes);
+
+  dgl_id_t *neg_dst_data = static_cast<dgl_id_t *>(neg_dst->data);
+  dgl_id_t *neg_src_data = static_cast<dgl_id_t *>(neg_src->data);
+  dgl_id_t *neg_eid_data = static_cast<dgl_id_t *>(neg_eid->data);
+  dgl_id_t *induced_neg_eid_data = static_cast<dgl_id_t *>(induced_neg_eid->data);
+
+  dgl_id_t curr_eid = 0;
+  std::vector<size_t> neg_vids;
+  neg_vids.reserve(chunk_size * 2);
+  std::unordered_map<dgl_id_t, dgl_id_t> neg_map;
+  for (int64_t i_chunk = 0; i_chunk < num_chunks; i_chunk++) {
+    size_t neg_idx = (chunk_size * 2 - 1) * chunk_size * 2 * i_chunk;
+    size_t pos_edge_idx = (chunk_size) * i_chunk;
+    neg_vids.clear();
+
+    std::vector<size_t> neighbors;
+    DGLIdIters neigh_it;
+    const dgl_id_t *unchanged;
+    const dgl_id_t *changed;
+    dgl_id_t *neg_unchanged;
+    dgl_id_t *neg_changed;
+
+    RandomSample(num_tot_nodes, chunk_size * 2, &neg_vids, &seed);
+    for (int tmp = 0; tmp != 2; ++tmp) {
+      // for each pos sample in chunk, need to corrupt head and tail
+      if (tmp == 0) {
+        unchanged = dst_data;
+        changed = src_data;
+        neg_unchanged = neg_dst_data;
+        neg_changed = neg_src_data;
+      } else {
+        unchanged = src_data;
+        changed = dst_data;
+        neg_unchanged = neg_src_data;
+        neg_changed = neg_dst_data;
+      }
+      for (int64_t in_chunk = 0; in_chunk != chunk_size; ++in_chunk) {
+        // first generate 2 * chunk_size by replacing head node and then tail
+        dgl_id_t global_unchanged = induced_vid_data[unchanged[pos_edge_idx + in_chunk]];
+        dgl_id_t local_unchanged = global2local_map(global_unchanged, &neg_map);
+        // generate 2 * chunk_size neg samples
+        for (int64_t j = 0; j != chunk_size * 2; ++j) {
```

Comment:
does PBG negative sampling always generate 2*chunk_size negative edges for each positive edge?